extensions/amp-analytics/0.1/events.js

```diff
@@ -402,6 +408,79 @@ export class IniLoadTracker extends EventTracker {
 
 
 /**
+ * Tracks video session events
+ */
+export class VideoEventTracker extends EventTracker {
+  /**
+   * @param {!./analytics-root.AnalyticsRoot} root
+   */
+  constructor(root) {
+    super(root);
+
+    /** @private @const {!Observable<!Event>} */
+    this.sessionObservable_ = new Observable();
+
+    /** @private @const */
+    this.boundOnSession_ = e => {
+      this.sessionObservable_.fire(e);
+    };
+
+    this.root.getRoot().addEventListener(
+        VideoEvents.ANALYTICS, this.boundOnSession_);
+  }
+
+  /** @override */
+  dispose() {
+    this.root.getRoot().removeEventListener(
+        VideoEvents.ANALYTICS, this.boundOnSession_);
+  }
+
+  /** @override */
+  add(context, eventType, config, listener) {
+    const videoSpec = config['videoSpec'] || {};
+    const endSessionWhenInvisible = videoSpec['end-session-when-invisible'];
+    const excludeAutoplay = videoSpec['exclude-autoplay'];
+    const selector = config['selector'] || videoSpec['selector'];
+    const selectionMethod = config['selectionMethod'] || null;
+    const on = config['on'];
+    const targetReady =
+        this.root.getElement(context, selector, selectionMethod);
+
+    return this.sessionObservable_.add(event => {
+      const eventData = getData(event);
+      const type = eventData['type'];
+      const details = eventData['details'];
+      const element = user().assertElement(event.target,
+          'No target specified by video session event.');
+
+      // Wait for target selected
+      targetReady.then(analyticsTarget => {
+        if (analyticsTarget.contains(element)) {
+          const isSessionVisible =
+              (type === VideoAnalyticsType.SESSION_VISIBLE);
+
+          // These spec flags filter events that are triggered.
+          // If they apply, we return and do not log the analytics events.
+          const filterSession = (isSessionVisible && !endSessionWhenInvisible);
```

Comment:
The point of the spec is to allow them filter events on the client-side instead of getting pinged for everything and then running custom logic to filter data later on their side. If their definition of ending a session is not based on visibility (e.g. news videos where audio is as important as the video itself) then getting a 'video-session` ping everytime the video goes in/out of view is just too much to deal with on their side to reconstruct what they consider a `video-session`